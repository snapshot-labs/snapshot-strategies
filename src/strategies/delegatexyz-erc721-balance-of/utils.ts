import { Address, DelegatedMapping, AddressScore } from './types';

export const lowerCaseAddress = (address: Address) =>
  address.toLowerCase() as Address;

export const calculateVotingPower = (
  inputAddresses: Address[],
  addressScores: AddressScore,
  walletMapping: DelegatedMapping
) => {
  const userVotingPower = {};
  inputAddresses.map((address: Address) => {
    let count = 0.0;
    walletMapping[lowerCaseAddress(address)].map((address: Address) => {
      count += addressScores[address];
    });
    userVotingPower[address] = count;
  });
  return userVotingPower;
};
