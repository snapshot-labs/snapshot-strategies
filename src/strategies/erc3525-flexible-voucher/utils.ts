const oneDaySeconds = 86400;

export const maturitiesCoefficient = (
  latestStartTime: number,
  terms: Array<any>
) => {
  const nowData = Date.parse(new Date().toString()) / 1000;
  const difference =
    latestStartTime + terms[terms.length - 1].toNumber() - nowData;

  if (difference <= 0) {
    return 1;
  } else if (difference > 0 && difference <= 90 * oneDaySeconds) {
    return 1.1;
  } else if (
    difference > 90 * oneDaySeconds &&
    difference <= 183 * oneDaySeconds
  ) {
    return 1.2;
  } else if (
    difference > 183 * oneDaySeconds &&
    difference <= 365 * oneDaySeconds
  ) {
    return 1.5;
  } else {
    return 2;
  }
};

export const claimCoefficient = (claimType: number) => {
  if (claimType == 0) {
    return 1.2;
  } else if (claimType == 1) {
    return 2;
  } else if (claimType == 2) {
    return 1.5;
  } else {
    return 1;
  }
};
