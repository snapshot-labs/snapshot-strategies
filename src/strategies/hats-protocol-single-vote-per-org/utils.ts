export function hatIdDecimalToHex(hatId) {
  return `0x` + BigInt(hatId).toString(16).padStart(64, `0`);
}

export function treeIdDecimalToHex(treeId) {
  return `0x` + treeId.toString(16).padStart(8, `0`);
}

export function hatIdHexToDecimal(hatId) {
  return BigInt(hatId);
}

export function treeIdHexToDecimal(treeId) {
  return parseInt(treeId, 16);
}

export function setCharAt(str, index, chr) {
  if (index > str.length - 1) return str;
  return str.substring(0, index) + chr + str.substring(index + 1);
}

export function hatIdDecimalToIp(hatId) {
  const hexId = hatIdDecimalToHex(hatId);
  let ip = treeIdHexToDecimal(hexId.substring(0, 10)).toString();
  for (let i = 10; i < hexId.length; i += 4) {
    const domainAtLevel = hexId.substring(i, i + 4);
    if (domainAtLevel === `0000`) {
      break;
    }
    ip += `.` + parseInt(domainAtLevel, 16);
  }
  return ip;
}
