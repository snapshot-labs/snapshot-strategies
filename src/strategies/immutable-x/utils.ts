import { Interface } from '@ethersproject/abi';
import { Contract } from '@ethersproject/contracts';

import { abi as multicallAbi } from './abi/Multicall.json';
import networks from './networks.json';

export async function multicall(network: string, provider, abi: any[], calls: any[], options?) {
    const multi = new Contract(networks[network].multicall, multicallAbi, provider);
    const itf = new Interface(abi);
    try {
        const [, res] = await multi.aggregate(
            calls.map((call) => [call[0].toLowerCase(), itf.encodeFunctionData(call[1], call[2])]),
            options || {}
        );
        return res.map((call, i) => itf.decodeFunctionResult(calls[i][1], call));
    } catch (e) {
        return Promise.reject(e);
    }
}
