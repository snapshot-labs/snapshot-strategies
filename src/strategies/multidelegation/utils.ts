import { getAddress } from '@ethersproject/address';
import {
  getDelegatesBySpace,
  getSnapshots,
  subgraphRequest
} from '../../utils';

export async function getPolygonDelegatesBySpace(
  subgraphUrl: string,
  space: string,
  snapshot = 'latest'
) {
  const spaceIn = ['', space];
  if (space.includes('.eth')) spaceIn.push(space.replace('.eth', ''));

  const PAGE_SIZE = 1000;
  let result: { delegator: string; space: string; delegate: string }[] = [];
  let page = 0;
  const params: any = {
    delegations: {
      __args: {
        where: {
          space_in: spaceIn
        },
        first: PAGE_SIZE,
        skip: 0
      },
      delegator: true,
      space: true,
      delegate: true
    }
  };
  if (snapshot !== 'latest') {
    params.delegations.__args.block = { number: Number(snapshot) };
  }

  while (true) {
    params.delegations.__args.skip = page * PAGE_SIZE;

    const pageResult = await subgraphRequest(subgraphUrl, params);
    const pageDelegations = pageResult.delegations || [];
    result = result.concat(pageDelegations);
    page++;
    if (pageDelegations.length < PAGE_SIZE) break;
  }

  return result;
}

export async function getMultiDelegations(
  subgraphUrl: string,
  space: string,
  snapshot?: string
): Promise<Map<string, string[]>> {
  const delegatesBySpace = await getPolygonDelegatesBySpace(
    subgraphUrl,
    space,
    snapshot
  );

  return delegatesBySpace.reduce((accum, delegation) => {
    const delegator = getAddress(delegation.delegator);
    const delegate = getAddress(delegation.delegate);
    const existingDelegates = accum.get(delegator) || [];
    accum.set(delegator, [...existingDelegates, delegate]);
    return accum;
  }, new Map<string, string[]>());
}

export async function getSingleDelegations(
  space: string,
  network: string,
  addresses: string[],
  snapshot: string
): Promise<Map<string, string>> {
  const delegatesBySpace = await getDelegatesBySpace(network, space, snapshot);
  const delegationsReverse = new Map<string, string>();

  delegatesBySpace.forEach((delegation: any) => {
    const delegator = delegation.delegator.toLowerCase();
    const delegate = delegation.delegate.toLowerCase();
    delegationsReverse.set(delegator, delegate);
  });

  delegatesBySpace
    .filter((delegation: any) => delegation.space !== '')
    .forEach((delegation: any) => {
      const delegator = delegation.delegator.toLowerCase();
      const delegate = delegation.delegate.toLowerCase();
      delegationsReverse.set(delegator, delegate);
    });

  const result = new Map<string, string>();
  addresses.forEach((address) => {
    const addressLc = address.toLowerCase();
    const delegate = delegationsReverse.get(addressLc);
    if (!!delegate) {
      const delegator = getAddress(addressLc);
      result.set(delegator, getAddress(delegate));
    }
  });

  return result;
}

// legacy and multi delegations are both objects with delegator as key and delegate(s) as value
export function mergeDelegations(
  legacyDelegations: Map<string, string>,
  multiDelegations: Map<string, string[]>
): Map<string, string[]> {
  const mergedDelegations: Map<string, string[]> = new Map();

  const delegators = new Set([
    ...(legacyDelegations?.keys() || []),
    ...(multiDelegations?.keys() || [])
  ]);

  for (const delegator of delegators) {
    const legacyDelegate = legacyDelegations.get(delegator);
    const multiDelegates = multiDelegations.get(delegator);

    if (multiDelegates && multiDelegates.length > 0) {
      mergedDelegations.set(delegator, multiDelegates);
    } else if (legacyDelegate) {
      mergedDelegations.set(delegator, [legacyDelegate]);
    }
  }

  return mergedDelegations;
}

// delegations is an object with delegator as key and delegate(s) as value
export function reverseDelegations(delegations: Map<string, string[]>) {
  const invertedDelegations = new Map<string, string[]>();

  for (const [delegator, delegates] of delegations) {
    for (const delegate of delegates) {
      if (invertedDelegations.has(delegate)) {
        invertedDelegations.get(delegate)?.push(delegator);
      } else {
        invertedDelegations.set(delegate, [delegator]);
      }
    }
  }

  return invertedDelegations;
}

export function getDelegatorScores(
  scores: Record<string, number>[],
  delegatorAddress: string
) {
  return scores.reduce((strategiesAcumScore, strategyScores) => {
    return strategyScores[delegatorAddress] !== undefined
      ? strategiesAcumScore + strategyScores[delegatorAddress]
      : strategiesAcumScore;
  }, 0);
}

export function getDelegateScore(
  delegations: string[],
  scores: Record<string, number>[],
  mergedDelegations: Map<string, string[]>
) {
  return delegations.reduce((delegationsAcumScore, delegatorAddress) => {
    const delegatorDelegations = mergedDelegations.get(delegatorAddress);
    const delegationsAmount = delegatorDelegations?.length || 1;
    return (
      delegationsAcumScore +
      getDelegatorScores(scores, delegatorAddress) / delegationsAmount
    );
  }, 0);
}

export function getAddressTotalDelegatedScore(
  delegate,
  mergedDelegations: Map<string, string[]>,
  reversedDelegations: Map<string, string[]>,
  scores
) {
  const delegations = reversedDelegations.get(delegate);
  const delegateScore = delegations
    ? getDelegateScore(delegations, scores, mergedDelegations)
    : 0;
  return [delegate, delegateScore];
}

export async function getPolygonBlockNumber(
  polygonChainId,
  network,
  snapshot,
  provider
) {
  const blocks = await getSnapshots(network, snapshot, provider, [
    polygonChainId
  ]);
  return blocks[polygonChainId];
}

export async function getPolygonMultiDelegations(
  multiDelegationEnv,
  network,
  snapshot,
  provider,
  delegationSpace
) {
  const polygonBlockNumber = await getPolygonBlockNumber(
    multiDelegationEnv.polygonChainId,
    network,
    snapshot,
    provider
  );

  return getMultiDelegations(
    multiDelegationEnv.subgraphUrl,
    delegationSpace,
    polygonBlockNumber
  );
}

export function getDelegationAddresses(
  reversedDelegations: Map<string, string[]>
) {
  return Array.from(
    Array.from(reversedDelegations.values()).reduce(
      (accumulator, addresses) => new Set([...accumulator, ...addresses]),
      new Set<string>()
    )
  );
}
