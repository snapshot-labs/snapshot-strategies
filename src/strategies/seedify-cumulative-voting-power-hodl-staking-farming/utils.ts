import { formatUnits } from '@ethersproject/units';

export const sfundStakingAbi = [
  'function userDeposits(address) external view returns (uint256, uint256, uint256, uint256, uint256, bool)'
];
export const farmingAbi = [
  'function userDeposits(address from) external view returns (uint256, uint256, uint256, uint256)'
];
export const bep20Abi = [
  'function totalSupply() view returns (uint256)',
  'function balanceOf(address) view returns (uint256)'
];

export const getStakingBalanceOf = (
  stakedBalances: any,
  userIndex: any,
  stakingAddrAmount: any,
  userAmount: any
) => {
  let sum = 0;
  let balance: any, userPosition: any;
  for (
    let stakingContractIndex = 0;
    stakingContractIndex < stakingAddrAmount;
    stakingContractIndex++
  ) {
    userPosition = userAmount * stakingContractIndex;
    balance = toDecimals(stakedBalances[userPosition + userIndex]['0']);
    sum += balance;
  }
  return sum;
};

export const toDecimals = (bigNumber: any) => {
  return parseFloat(formatUnits(bigNumber.toString(), 18));
};

export const calculateBep20InLPForUser = (
  lpStaked: any,
  totalLPSupply: any,
  totalBep20InPool: any
) => {
  lpStaked = toDecimals(lpStaked['0']);

  return (lpStaked / totalLPSupply) * totalBep20InPool;
};

export const createCallsToReadUsersData: any = (
  addresses: any[],
  stakingAddresses: any[],
  functionToCall: string
) => {
  const promises: any = [];
  for (let i = 0; i < stakingAddresses.length; i++) {
    promises.push(
      ...createCallToReadUsersData(
        addresses,
        stakingAddresses[i],
        functionToCall
      )
    );
  }
  return promises;
};

export const createCallToReadUsersData: any = (
  addresses: any,
  contractAddress: any,
  functionToCall: any
) => {
  return addresses.map((address: any) => [
    contractAddress,
    functionToCall,
    [address]
  ]);
};
