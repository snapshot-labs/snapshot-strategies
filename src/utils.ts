import _strategies from './strategies';
import snapshot from '@snapshot-labs/snapshot.js';

export async function getScoresDirect(
  space: string,
  strategies: any[],
  network: string,
  provider,
  addresses: string[],
  snapshot: number | string = 'latest'
) {
  try {
    return await Promise.all(
      strategies.map((strategy) =>
        (snapshot !== 'latest' && strategy.params?.start > snapshot) ||
        (strategy.params?.end &&
          (snapshot === 'latest' || snapshot > strategy.params?.end)) ||
        addresses.length === 0
          ? {}
          : _strategies[strategy.name].strategy(
              space,
              network,
              provider,
              addresses,
              strategy.params,
              snapshot
            )
      )
    );
  } catch (e) {
    return Promise.reject(e);
  }
}

export const {
  multicall,
  Multicaller,
  subgraphRequest,
  ipfsGet,
  call,
  getBlockNumber,
  getProvider
} = snapshot.utils;

export default {
  getScoresDirect,
  multicall,
  Multicaller,
  subgraphRequest,
  ipfsGet,
  call,
  getBlockNumber,
  getProvider
};
