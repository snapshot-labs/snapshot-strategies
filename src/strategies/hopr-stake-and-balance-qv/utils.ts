import { BigNumber } from '@ethersproject/bignumber';
import { subgraphRequest } from '../../utils';
import { parseUnits } from '@ethersproject/units';

/*
 ******************************************
 ****************** TYPES *****************
 ******************************************
 */
// details of a safe created by the NodeSafeFactory contract
export type Safe = {
  safeAddress: string;
  owners: string[];
  nodes: string[];
};

/*
 ******************************************
 *************** PARAMETERS ***************
 ******************************************
 */
const QUERY_LIMIT = 1000; // 1000 addresses per query in Subgraph

/*
 ***********************************************
 **************** SUBGRAPH SETUP ***************
 ***********************************************
 */
export function getStudioProdSubgraphUrl(
  apiKey: string | null | undefined,
  subgraphId: string
): string | null {
  return !apiKey
    ? null
    : `https://gateway.thegraph.com/api/${apiKey}/subgraphs/id/${subgraphId}`;
}

export function getStudioDevSubgraphUrl(
  accountStudioId: string | null | undefined,
  subgraphName: string,
  version: string
): string | null {
  return !accountStudioId
    ? null
    : `https://api.studio.thegraph.com/query/${accountStudioId}/${subgraphName}/${version}`;
}

export function getHostedSubgraphUrl(
  accountName: string,
  subgraphName: string | null
): string | null {
  return !subgraphName
    ? null
    : `https://api.thegraph.com/subgraphs/name/${accountName}/${subgraphName}`;
}

/**
 * Try to query subgraphs from three differnt endpoints (hosted service, studio for development, studio in production), if applicable
 * @param hostedSubgraphUrl hosted subgrpah url
 * @param stuidoDevSubgraphUrl development url foro studio subgraph
 * @param studioProdSubgraphUrl production url foro studio subgraph
 * @param builtQuery query object
 * @returns null or an object of summed token balance per address
 */
export async function subgraphRequestsToVariousServices(
  hostedSubgraphUrl: string | null,
  stuidoDevSubgraphUrl: string | null,
  studioProdSubgraphUrl: string | null,
  builtQuery: any
): Promise<any> {
  if (hostedSubgraphUrl) {
    try {
      // first try with hosted service
      return subgraphRequest(hostedSubgraphUrl, builtQuery);
    } catch (error) {
      // console.log('Failed to get data from hostedSubgraphUrl');
    }
  }

  // then try with studio dev service
  if (stuidoDevSubgraphUrl) {
    try {
      return subgraphRequest(stuidoDevSubgraphUrl, builtQuery);
    } catch (error) {
      // console.log('Failed to get data from stuidoDevSubgraphUrl');
    }
  }

  // then try with studio prod service
  if (studioProdSubgraphUrl) {
    try {
      return subgraphRequest(studioProdSubgraphUrl, builtQuery);
    } catch (error) {
      // console.log('Failed to get data from studioProdSubgraphUrl');
    }
  }
  return null;
}

/*
 *************************************************
 **************** SUBGRAPH QUERIES ***************
 *************************************************
 */
/**
 * Get block number from Gnosis chain at a given timestamp.
 * The timestamp of the returned block should be no-bigger than the desired timestamp
 * @param queryUrl URL to the subgraph query URL
 * @param timestamp number of timestamp
 * @param fallbackBlockNumber fallback block number on Gnosis chain, in case no result gets returned.
 * @returns a number
 */
export async function getGnosisBlockNumber(
  queryUrl: string,
  timestamp: number,
  fallbackBlockNumber: number
): Promise<number> {
  const query = {
    blocks: {
      __args: {
        first: 1,
        orderBy: 'number',
        orderDirection: 'desc',
        where: {
          timestamp_lte: timestamp
        }
      },
      number: true,
      timestamp: true
    }
  };

  // query from subgraph
  try {
    const data = await subgraphRequestsToVariousServices(
      queryUrl,
      null,
      null,
      query
    );
    return !data ? fallbackBlockNumber : Number(data.blocks[0].number);
  } catch (error) {
    // console.error(`cannot convert block number from subgraph due to ${error}`);
    return fallbackBlockNumber;
  }
}

/**
 * Get the list of safe address created by the HoprStakeFactory contract
 * where the voting account is an owner.
 * It also returns the share per owner (1 / total number of owners) of each safe.
 * @param hostedSubgraphUrl url to the hosted subgraph
 * @param stuidoDevSubgraphUrl url to the dev subgraph in the studio
 * @param studioProdSubgraphUrl url to the production subgraph in the studio
 * @param addresses address of voting accounts, which is an owner of the safe
 * @param blockNumber Gnosis chain block number for the query
 * @returns a key-value object where the key is safe address the value is the total number of owners.
 */
export async function safeStakeSubgraphQuery(
  hostedSubgraphUrl: string | null,
  stuidoDevSubgraphUrl: string | null,
  studioProdSubgraphUrl: string | null,
  addresses: string[],
  blockNumber: number
): Promise<Safe[]> {
  const query = {
    safes: {
      __args: {
        first: QUERY_LIMIT,
        block: { number: blockNumber },
        where: {
          owners_: {
            owner_in: addresses.map((adr) => adr.toLowerCase())
          }
        }
      },
      id: true,
      owners: {
        owner: {
          id: true
        }
      },
      registeredNodesInNetworkRegistry: {
        node: {
          id: true
        }
      }
    }
  };

  // query from subgraph
  const data = await subgraphRequestsToVariousServices(
    hostedSubgraphUrl,
    stuidoDevSubgraphUrl,
    studioProdSubgraphUrl,
    query
  );

  // return parsed entries
  if (!data || !data.safes || data.safes.length == 0) {
    return [];
  } else {
    return data.safes.map((s) => {
      return {
        safeAddress: s.id,
        owners: s.owners.map((o) => o.owner.id),
        nodes: s.registeredNodesInNetworkRegistry.map((n) => n.node.id)
      } as Safe;
    });
  }
}

/**
 * Get the list of wxHOPR + xHOPR balance of addresses on Gnosis chain
 * @param hostedSubgraphUrl url to the hosted subgraph
 * @param stuidoDevSubgraphUrl url to the dev subgraph in the studio
 * @param studioProdSubgraphUrl url to the production subgraph in the studio
 * @param addresses address of wallets
 * @param blockNumber Gnosis chain block number for the query
 * @returns a key-value object where the key is the address and the value is the total HOPR token balance on Gnosis chain.
 */
export async function hoprTotalOnGnosisSubgraphQuery(
  hostedSubgraphUrl: string | null,
  stuidoDevSubgraphUrl: string | null,
  studioProdSubgraphUrl: string | null,
  addresses: string[],
  blockNumber: number
): Promise<{ [propName: string]: BigNumber }> {
  const query = {
    accounts: {
      __args: {
        first: QUERY_LIMIT,
        block: { number: blockNumber },
        where: {
          id_in: addresses.map((adr) => adr.toLowerCase())
        }
      },
      id: true,
      totalBalance: true
    }
  };

  // query from subgraph
  const data = await subgraphRequestsToVariousServices(
    hostedSubgraphUrl,
    stuidoDevSubgraphUrl,
    studioProdSubgraphUrl,
    query
  );

  // map result (data.accounts) to addresses
  const entries = !data
    ? addresses.map((addr) => [addr, BigNumber.from('0')])
    : data.accounts.map((d) => [
        d.id,
        parseUnits(d.totalBalance.toString(), 18)
      ]);
  return Object.fromEntries(entries);
}

/**
 * Get the total stake in all the outgoing channels per node
 * @param hostedSubgraphUrl url to the hosted subgraph
 * @param stuidoDevSubgraphUrl url to the dev subgraph in the studio
 * @param studioProdSubgraphUrl url to the production subgraph in the studio
 * @param addresses node addresses
 * @param blockNumber Gnosis chain block number for the query
 * @returns a key-value object where the key is the address and the value is the total HOPR token balance on Gnosis chain.
 */
export async function hoprNodeStakeOnChannelsSubgraphQuery(
  hostedSubgraphUrl: string | null,
  stuidoDevSubgraphUrl: string | null,
  studioProdSubgraphUrl: string | null,
  addresses: string[],
  blockNumber: number
): Promise<{ [propName: string]: BigNumber }> {
  const query = {
    accounts: {
      __args: {
        first: QUERY_LIMIT,
        block: { number: blockNumber },
        where: {
          id_in: addresses.map((adr) => adr.toLowerCase())
        }
      },
      id: true,
      balance: true
    }
  };

  // query from subgraph
  const data = await subgraphRequestsToVariousServices(
    hostedSubgraphUrl,
    stuidoDevSubgraphUrl,
    studioProdSubgraphUrl,
    query
  );

  // map result (data.accounts) to addresses
  const entries = !data
    ? addresses.map((addr) => [addr, BigNumber.from('0')])
    : data.accounts.map((d) => [d.id, parseUnits(d.balance.toString(), 18)]);
  return Object.fromEntries(entries);
}

/*
 ***********************************************
 ******************** OTHERS *******************
 ***********************************************
 */
export function trimArray<T>(
  originalArray: Array<T>,
  size: number = QUERY_LIMIT
): Array<Array<T>> {
  return Array.apply(null, Array(Math.ceil(originalArray.length / size))).map(
    (_e, i) => originalArray.slice(i * size, (i + 1) * size)
  );
}
