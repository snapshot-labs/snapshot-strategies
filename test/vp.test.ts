import { getVp } from '../src/utils/vp';

const address = '0xeF8305E140ac520225DAf050e2f71d5fBcC543e7';
const space = 'cvx.eth';
const network = '1';
const snapshot = 15354134;
const strategies = [
  {
    name: 'erc20-balance-of',
    params: {
      symbol: 'CVX',
      address: '0x72a19342e8F1838460eBFCCEf09F6585e32db86E',
      decimals: 18
    }
  },
  {
    name: 'eth-balance',
    network: '100',
    params: {}
  },
  {
    name: 'eth-balance',
    network: '1',
    params: {}
  },
  {
    name: 'eth-balance',
    network: '10',
    params: {}
  }
];

describe('', () => {
  it('getVp without delegation', async () => {
    const scores = await getVp(
      address,
      network,
      strategies,
      snapshot,
      space,
      false
    );
    expect(scores).toMatchSnapshot();
    console.log(scores);
  }, 20e3);

  it('getVp with delegation', async () => {
    const scores = await getVp(
      address,
      network,
      strategies,
      snapshot,
      space,
      true
    );
    expect(scores).toMatchSnapshot();
    console.log(scores);
  }, 20e3);
});
