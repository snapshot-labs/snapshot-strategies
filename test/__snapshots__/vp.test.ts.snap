// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` getVp with delegation 1`] = `
Object {
  "vp": 10.077130335431244,
  "vp_by_strategy": Array [
    0,
    10.028706352441185,
    0,
    0.04842398299005922,
  ],
  "vp_state": "final",
}
`;

exports[` getVp without delegation 1`] = `
Object {
  "vp": 21.55404462002206,
  "vp_by_strategy": Array [
    0,
    9.998985610441185,
    11.506635026590818,
    0.04842398299005922,
  ],
  "vp_state": "final",
}
`;
